index.d.ts

```diff
@@ -1,74 +1,91 @@
 
-  interface Position {
-    readonly start: Point
-    readonly end: Point
-    readonly indent?: number[]
-  }
+  type Position = unist.Position
 
-  export type NodeType = Content['type']
+  type NodeType = mdast.Content['type']
 
-  export type AlignType = 'left' | 'right' | 'center' | null
+  type AlignType = mdast.AlignType
 
-  export type ReferenceType = 'shortcut' | 'collapsed' | 'full'
+  type ReferenceType = mdast.ReferenceType
 
-  export type LinkTargetResolver = (uri: string, text: string, title?: string) => string
+  type LinkTargetResolver = (uri: string, text: string, title?: string) => string
 
-  type Renderer<T> = (props: T) => ReactElement<T>
+  type Renderer<T> = (props: T) => ElementType<T>
   interface Renderers {
     [key: string]: string | Renderer<any>
   }
 
-  export interface ReactMarkdownProps {
+  interface ReactMarkdownPropsBase {
     readonly className?: string
-    readonly source?: string
-    readonly children?: string
     readonly sourcePos?: boolean
     readonly includeNodeIndex?: boolean
     readonly rawSourcePos?: boolean
-    readonly escapeHtml?: boolean
-    readonly allowDangerousHtml?: boolean
-    readonly skipHtml?: boolean
-    readonly allowNode?: (node: Content, index: number, parent: NodeType) => boolean
-    readonly allowedTypes?: NodeType[]
-    readonly disallowedTypes?: NodeType[]
+    readonly allowNode?: (node: mdast.Content, index: number, parent: NodeType) => boolean
     readonly linkTarget?: string | LinkTargetResolver
     readonly transformLinkUri?:
       | ((uri: string, children?: ReactNode, title?: string) => string)
       | null
     readonly transformImageUri?:
       | ((uri: string, children?: ReactNode, title?: string, alt?: string) => string)
       | null
-    readonly unwrapDisallowed?: boolean
-    readonly renderers?: {[nodeType: string]: ReactType}
+    readonly renderers?: {[nodeType: string]: ElementType}
+    /** @deprecated use plugins */
     readonly astPlugins?: PluggableList
     readonly plugins?: PluggableList
   }
 
-  export var types: NodeType[]
-  export var renderers: Renderers
-  export var uriTransformer: (uri: string) => string
+  interface SourceProp {
+    /** @deprecated use children */
+    readonly source: string
+  }
+
+  interface ChildrenProp {
+    readonly children: string
+  }
+
+  interface AllowedTypesProp {
+    readonly allowedTypes?: NodeType[]
+  }
+
+  interface DisallowedTypesProp {
+    readonly disallowedTypes: NodeType[]
+    readonly unwrapDisallowed?: boolean
+  }
+
+  interface EscapeHtmlProp {
+    /** @deprecated use allowDangerousHtml */
+    readonly escapeHtml?: boolean
+    readonly allowDangerousHtml?: boolean
+  }
+
+  interface SkipHtmlProp {
+    readonly skipHtml?: boolean
+  }
+
+  type ReactMarkdownProps = ReactMarkdownPropsBase &
+    MutuallyExclusive<ChildrenProp, SourceProp> &
+    MutuallyExclusive<AllowedTypesProp, DisallowedTypesProp> &
+    MutuallyExclusive<EscapeHtmlProp, SkipHtmlProp>
```

Comment:
Prevents exclusive types from overlapping.
It's a bit verbose, but may be needed until Exact types microsoft/TypeScript#12936 become a thing